### PR TITLE
Explicit field hierarchy and array sub-types

### PR DESCRIPTION
**Pull request by richfab**
Initial PR: https://github.com/MobilityData/gbfs/pull/576

## Problem
- The field hierarchy with dashes is difficult to read.
- The description of required fields inside optional objects is too verbose.
- The type Array is vague.

## Solution
This PR makes the field hierarchy and the array sub-types explicit:

Before | After
-- | --
image | Screenshot 2024-01-04 at 11 02 55

The changes are editorial only. A vote is not required.

### CODE REVIEW

File: gbfs.md [hunk 1]
```diff
@@ -1349,7 +1348,7 @@ A `Rule` object defines the set of restrictions in place for a particular zone.
 
 Field Name | REQUIRED | Type | Defines
 ---|---|---|---
-`vehicle_type_ids` <br/>*(as of v3.0-RC2)* | OPTIONAL | Array | Array of IDs of vehicle types for which any restrictions SHOULD be applied (see vehicle type definitions in `vehicle_types.json`). If vehicle type IDs are not specified, then restrictions apply to all vehicle types.
+`vehicle_type_ids` <br/>*(as of v3.0-RC2)* | OPTIONAL | Array&lt;String&gt; | Array of IDs of vehicle types for which any restrictions SHOULD be applied (see vehicle type definitions in `vehicle_types.json`). If vehicle type IDs are not specified, then restrictions apply to all vehicle types.
```
Comment (tdelmas) [inline gbfs.md]: Array&lt;ID&gt;

File: gbfs.md [hunk 2]
```diff
@@ -552,7 +552,7 @@ Field Name | REQUIRED | Type | Defines
...
-`vehicle_types[].pricing_plan_ids`<br/>*(added in v2.3)* | OPTIONAL | Array | Array of all pricing plan IDs, as defined in `system_pricing_plans.json`, that are applied to this vehicle type. <br /><br />This array SHOULD be published when there are multiple pricing plans defined in `system_pricing_plans.json` that apply to a single vehicle type.
+`vehicle_types[].pricing_plan_ids`<br/>*(added in v2.3)* | OPTIONAL | Array&lt;String&gt; | Array of all pricing plan IDs, as defined in `system_pricing_plans.json`, that are applied to this vehicle type. <br /><br />This array SHOULD be published when there are multiple pricing plans defined in `system_pricing_plans.json` that apply to a single vehicle type.
```
Comment (tdelmas) [inline gbfs.md]: Array&lt;ID&gt;

File: gbfs.md [hunk 3]
```diff
@@ -717,18 +717,18 @@ Field Name | REQUIRED | Type | Defines
...
 `stations[].vehicle_types_capacity` <br/>*(as of v3.0-RC2)* | OPTIONAL | Array&lt;Object&gt; | These objects are used to model the parking capacity of virtual stations (defined using the `is_virtual_station` field) for each vehicle type that can be returned to this station. The total number of vehicles from each of these objects SHOULD add up to match the value specified in the `capacity` field.
-`stations[].vehicle_types_capacity[].vehicle_type_ids` <br/>*(as of v3.0-RC2)* | REQUIRED | Array | The `vehicle_type_ids`, as defined in `vehicle_types.json`, that may park at the virtual station.
+`stations[].vehicle_types_capacity[].vehicle_type_ids` <br/>*(as of v3.0-RC2)* | REQUIRED | Array&lt;String&gt; | The `vehicle_type_ids`, as defined in `vehicle_types.json`, that may park at the virtual station.
```
Comment (tdelmas) [inline gbfs.md]: Array&lt;ID&gt;

File: gbfs.md [hunk 4]
```diff
@@ -717,18 +717,18 @@ Field Name | REQUIRED | Type | Defines
...
 `stations[].vehicle_docks_capacity` <br/>*(as of v3.0-RC2)* | OPTIONAL | Array&lt;Object&gt; | These objects are used to model the total docking capacity of a station, both available and unavailable, for each type of vehicle that may dock at this station. The total number of docks from each of these objects SHOULD add up to match the value specified in the `capacity` field.
-`stations[].vehicle_docks_capacity[].vehicle_type_ids` <br/>*(as of v3.0-RC2)* | REQUIRED | Array | An array of strings where each string represents a `vehicle_type_id` that is able to use a particular type of dock at the station.
+`stations[].vehicle_docks_capacity[].vehicle_type_ids` <br/>*(as of v3.0-RC2)* | REQUIRED | Array&lt;String&gt; | An array of strings where each string represents a `vehicle_type_id` that is able to use a particular type of dock at the station.
```
Comment (tdelmas) [inline gbfs.md]: Array&lt;ID&gt;

File: gbfs.md [hunk 5]
```diff
@@ -851,12 +851,12 @@ Field Name | REQUIRED | Type | Defines
...
+`stations[].vehicle_docks_available` <br/>*(added in v2.1)* | Conditionally REQUIRED | Array&lt;Object&gt; | This field is REQUIRED in feeds where the [vehicle_types.json](#vehicle_typesjson) is defined and where certain docks are only able to accept certain vehicle types. If every dock at the station is able to accept any vehicle type, then this field is not REQUIRED. This field's value is an array of objects. Each of these objects is used to model the number of docks available for certain vehicle types. The total number of docks from each of these objects SHOULD add up to match the value specified in the `num_docks_available` field.
 `stations[].vehicle_docks_available[].vehicle_type_ids` <br/>*(added in v2.1)* | REQUIRED | Array&lt;String&gt; | Each string represents a `vehicle_type_id` that is able to use a particular type of dock at the station
```
Comment (tdelmas) [inline gbfs.md]: Array&lt;ID&gt;

File: gbfs.md [hunk 6]
```diff
@@ -1261,8 +1261,8 @@ Field Name | REQUIRED | Type | Defines
 `alerts[].times` | OPTIONAL | Array&lt;Object&gt; | The fields `start` and `end` indicate when the alert is in effect (for example, when the system or station is actually closed, or when a station is scheduled to be moved).
 `alerts[].times[].start` | REQUIRED | Timestamp | Start time of the alert.
 `alerts[].times[].end` | OPTIONAL | Timestamp | End time of the alert. If there is currently no end time planned for the alert, this can be omitted.
-`alerts[].station_ids` | OPTIONAL | Array | If this is an alert that affects one or more stations, include their ID(s). Otherwise omit this field. If both `station_ids` and `region_ids` are omitted, this alert affects the entire system.
-`alerts[].region_ids` | OPTIONAL | Array | If this system has regions, and if this alert only affects certain regions, include their ID(s). Otherwise, omit this field. If both `station_ids` and `region_ids` are omitted, this alert affects the entire system.
+`alerts[].station_ids` | OPTIONAL | Array&lt;String&gt; | If this is an alert that affects one or more stations, include their ID(s). Otherwise omit this field. If both `station_ids` and `region_ids` are omitted, this alert affects the entire system.
```
Comment (tdelmas) [inline gbfs.md]: Array&lt;ID&gt;

File: gbfs.md [hunk 7]
```diff
@@ -1261,8 +1261,8 @@ Field Name | REQUIRED | Type | Defines
...
+`alerts[].region_ids` | OPTIONAL | Array&lt;String&gt; | If this system has regions, and if this alert only affects certain regions, include their ID(s). Otherwise, omit this field. If both `station_ids` and `region_ids` are omitted, this alert affects the entire system.
```
Comment (tdelmas) [inline gbfs.md]: Array&lt;ID&gt;

File: gbfs.md [hunk 8]
```diff
@@ -1348,7 +1348,7 @@ A `Rule` object defines the set of restrictions in place for a particular zone.
 
 Field Name | REQUIRED | Type | Defines
 ---|---|---|---
-`vehicle_type_ids` <br/>*(as of v3.0-RC2)* | OPTIONAL | Array | Array of IDs of vehicle types for which any restrictions SHOULD be applied (see vehicle type definitions in `vehicle_types.json`). If vehicle type IDs are not specified, then restrictions apply to all vehicle types.
+`vehicle_type_ids` <br/>*(as of v3.0-RC2)* | OPTIONAL | Array&lt;String&gt; | Array of IDs of vehicle types for which any restrictions SHOULD be applied (see vehicle type definitions in `vehicle_types.json`). If vehicle type IDs are not specified, then restrictions apply to all vehicle types.
```
Comment (tdelmas) [inline gbfs.md]: Array&lt;ID&gt;

**Review by tdelmas — CHANGES_REQUESTED** [review]
Arrays of IDs should be defined as such. (descriptions may need to be updated)

**Comment by richfab** [comment]
Great catch for `Array<ID>` @tdelmas! Thanks!

**Review by tdelmas — APPROVED** [review]

File: gbfs.md [hunk 9]
```diff
@@ -188,7 +188,7 @@ The data returned by the near-realtime endpoints `station_status.json` and `vehi
 It is RECOMMENDED that all GBFS data sets be offered under an open data license. Open data licenses allow consumers to freely use, modify, and share GBFS data for any purpose in perpetuity. Licensing of GBFS data provides certainty to GBFS consumers, allowing them to integrate GBFS data into their work. All GBFS data sets SHOULD specify a license using the `license_id` field with an [SPDX identifier](https://spdx.org/licenses/) or by using the `license_url` field with a URL pointing to a custom license in `system_information.json`. See the GBFS repo for a [comparison of a subset of standard licenses](https://github.com/MobilityData/gbfs/blob/master/data-licenses.md).
 
 ## Field Types
-
+* Object - A JSON element consisting of key-value pairs (fields).
```
Comment (josee-sabourin) [inline gbfs.md]: I would move this back in between Non-negative Integer and Phone Number to maintain the alphabetized list

File: gbfs.md [hunk 10]
```diff
@@ -425,23 +424,23 @@ Field Name | REQUIRED | Type | Defines
...
+`brand_assets.brand_last_modified` <br/>*(added in v2.3)*  | REQUIRED | Date | Date that indicates the last time any included brand assets were updated or modified.
+`brand_assets.brand_terms_url` <br/>*(added in v2.3)*   | OPTIONAL |  URL |  A fully qualified URL pointing to the location of a page that defines the license terms of brand icons, colors, or other trademark information.  This field MUST NOT take the place of `license_url` or `license_id`.
+`brand_assets.brand_image_url` <br/>*(added in v2.3)*  |  REQUIRED |  URL | A fully qualified URL pointing to the location of a graphic file representing the brand for the service. File MUST be in SVG V1.1 format and MUST be either square or round.
+`brand_assets.brand_image_url_dark` <br/>*(added in v2.3)*  | OPTIONAL |  URL | A fully qualified URL pointing to the location of a graphic file representing the brand for the service for use in dark mode applications.  File MUST be in SVG V1.1 format and MUST be either square or round.
+`brand_assets.color` <br/>*(added in v2.3)*  | OPTIONAL |  String |  Color used to represent the brand for the service expressed as a 6 digit hexadecimal color code in the form #000000.
```
Comment (josee-sabourin) [inline gbfs.md:432]: I am not 100% sure but shouldn't there be a [ ] in these field names?

File: gbfs.md [hunk 11]
```diff
@@ -425,23 +424,23 @@ Field Name | REQUIRED | Type | Defines
...
+`rental_apps.ios.store_uri`  | REQUIRED | URI | URI where the rental iOS app can be downloaded from. Typically this will be a URI to an app store, such as the Apple App Store. If the URI points to an app store, the URI SHOULD follow iOS best practices so the viewing app can directly open the URI to the native app store app instead of a website.<br><br>See the [Analytics](#analytics) section for how viewing apps can report the origin of the deep link to rental apps. <br><br>Example value: `https://apps.apple.com/app/apple-store/id123456789`
+`rental_apps.ios.discovery_uri`  | REQUIRED | URI | URI that can be used to discover if the rental iOS app is installed on the device (for example, using [`UIApplication canOpenURL:`](https://developer.apple.com/documentation/uikit/uiapplication/1622952-canopenurl?language=objc)). This intent is used by viewing apps to prioritize rental apps for a particular user based on whether they already have a particular rental app installed.<br><br>Example value: `com.example.ios://`
```
Comment (josee-sabourin) [inline gbfs.md:443]: same comment as above

File: gbfs.md [hunk 12]
```diff
@@ -948,25 +947,25 @@ Field Name | REQUIRED | Type | Defines
...
+`vehicles[].rental_uris.web` | OPTIONAL | URL | URL that can be used by a web browser to show more information about renting a vehicle at this vehicle. <br><br>This URL SHOULD be a deep link specific to this vehicle, and SHOULD NOT be a general rental page that includes information for more than one vehicle.  The deep link SHOULD take users directly to this vehicle, without any prompts, interstitial pages, or logins. Make sure that users can see this vehicle even if they never previously opened the application. Note that providers MUST rotate identifiers within deep links after each rental to avoid unintentionally exposing private vehicle trip origins and destinations.<br><br>If this field is empty, it means deep linking is not supported for web browsers. <br><br>Example value: `https://www.example.com/app?sid=1234567890`
```
Comment (josee-sabourin) [inline gbfs.md:959]: and here

File: gbfs.md [hunk 13]
```diff
@@ -707,36 +706,36 @@ All stations included in `station_information.json` are considered public (meani
...
+`stations[].rental_uris.web` | OPTIONAL | URL | URL that can be used by a web browser to show more information about renting a vehicle at this station. <br><br>This URL SHOULD be a deep link specific to this station, and SHOULD NOT be a general rental page that includes information for more than one station.  The deep link SHOULD take users directly to this station, without any prompts, interstitial pages, or logins. Make sure that users can see this station even if they never previously opened the application.  <br><br>If this field is empty, it means deep linking is not supported for web browsers. <br><br>Example value: `https://www.example.com/app?sid=1234567890`
```
Comment (josee-sabourin) [inline gbfs.md:738]: here as well

File: gbfs.md (hunk 10 above)
Comment (richfab) [reply gbfs.md:432]: Good question! 
[ ] applies to arrays only. `brand_assets` is an object so that's why there is no [ ].

File: gbfs.md (hunk 11 above)
Comment (richfab) [reply gbfs.md:443]: `rental_apps` is not an array, so no need for [ ].

File: gbfs.md (hunk 12 above)
Comment (richfab) [reply gbfs.md:959]: `rental_uris` is not an array, so no need for [ ].

File: gbfs.md (hunk 13 above)
Comment (richfab) [reply gbfs.md:738]: rental_uris is not an array, so no need for [ ].

**Review by josee-sabourin — APPROVED** [review]

File: gbfs.md (hunk 13 above)
Comment (futuretap) [reply gbfs.md:738]: The sentence `Please use iOS Universal Links if possible so viewing apps do not need to manually manage the redirect of the user to the app store if the user does not have the application installed` is not correct. Universal Links do not fallback to the App Store if the app is not installed. Instead they open the website in the browser. That website in turn could redirect to the App Store but this is not a feature of Universal Links per se. So I would write: "Please use iOS Universal Links if possible so viewing apps provide a fallback web experience if the application is not installed."

Comment (tdelmas) [reply gbfs.md:738]: @futuretap I think you should open an issue about iOS links

Comment (futuretap) [reply gbfs.md:738]: @tdelmas done: https://github.com/MobilityData/gbfs/pull/585

**Comment by mplsmitch** [comment]
Hello @josee-sabourin @tdelmas -   I'm wondering if you can tell me anything about what's in this PR that you reviewed for Fabien. Someone brought it to my attention that there are a number of fields in v3.0 that are REQUIRED when they should be Conditionally REQUIRED.  In v2.3 everything is correct., the changes happened in this PR. For example if you look at `brand_assets.brand_last_modified` and it's description in this PR, it's been changed from v2.3. Fabien is away on leave otherwise I would ask him.  Thanks!

**Comment by josee-sabourin** [comment]
IIRC it would likely have to do with this: 
Screenshot 2025-12-16 at 3 22 15 PM

Anything that was Conditionally Required that was inside an **optional** object was changed to be required 

AKA if the object is optional and not included in the feed, you would logically not need to include the fields inside the object

i agree this is confusing, I'm sure at the time there was a catalyst for this but I can't remember

**Comment by mplsmitch** [comment]
Thank you @josee-sabourin !  You're right, I found the original discussion in #576.  The logic all checks out but it only if you've read the `REQUIRED` term definition at the top.

**Pull request merged**